### PR TITLE
Ahd demosaic speedup

### PR DESCRIPTION
**Pull request by heckflosse**

### CODE REVIEW

File: rtengine/ahd_demosaic_RT.cc [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+    constexpr double xyz_rgb[3][3] = {        /* XYZ from RGB */
+        { 0.412453, 0.357580, 0.180423 },
+        { 0.212671, 0.715160, 0.072169 },
+        { 0.019334, 0.119193, 0.950227 }
+    };
+
+    constexpr float d65_white[3] = { 0.950456, 1, 1.088754 };
+
+    volatile double progress = 0.0;
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: Don't know what you are trying to achieve with `volatile`. I can only only guess, it's some kind of thread safety. You're using `progress` in a critial section. Did you have problems with instruction reordering by the compiler? I think OMP mandates for reordering barriers in the right spot, but you're the expert.

Bottom line: There are only a few use cases for `volatile`, and I don't see one here. I hope you can enlighten me. :wink:

File: rtengine/ahd_demosaic_RT.cc [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    if (plistener) {
+        plistener->setProgressStr (Glib::ustring::compose(M("TP_RAW_DMETHOD_PROGRESSBAR"), RAWParams::BayerSensor::getMethodString(RAWParams::BayerSensor::Method::AHD)));
+        plistener->setProgress (0.0);
+    }
+
+    for (int i = 0; i < 0x10000; i++) {
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: `0x10000` -> `65536` for readability, as there's `65535.0` in the line below.

File: rtengine/ahd_demosaic_RT.cc [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        double r = (double)i / 65535.0;
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: `double` has a higher priority in coercion than `int`, so you can drop the ugly `(double)` C-style cast. As a bonus, `r` could be `const`.

File: rtengine/ahd_demosaic_RT.cc [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+#ifdef _OPENMP
+#pragma omp parallel
+#endif
+{
+    int progresscounter = 0;
+    float (*pix), (*rix)[3];
+    float ldiff[2][4], abdiff[2][4];
+    float (*lix)[3];
+    char *buffer = (char *) malloc (12 * TS * TS * sizeof(float) + 2 * TS * TS * sizeof(uint16_t)); /* 1053 kB per core */
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: A raw `malloc()` in 2018? :grin:

With `<cstdint>` `uint16_t` is now part of namespace `std`.

File: rtengine/ahd_demosaic_RT.cc [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                            rix[0][2 - cng] = CLIP(pix[0] + (0.5f * (pix[-1] + pix[1]
+                                                       - rix[-1][1] - rix[1][1] ) ));
+                            rix[0][cng] = CLIP(pix[0] + (0.5f * (pix[-width] + pix[width]
+                                                       - rix[-TS][1] - rix[TS][1])));
+                            rix[0][1] = pix[0];
+                        } else {
+                            rix[0][cng] = CLIP(rix[0][1] + (0.25f * (pix[-width - 1] + pix[-width + 1]
+                                                        + pix[+width - 1] + pix[+width + 1]
+                                                        - rix[-TS - 1][1] - rix[-TS + 1][1]
+                                                        - rix[+TS - 1][1] - rix[+TS + 1][1])));
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc:123]: Didn't check this (and the above) relative indexing dance. I assume, you've double-checked it...

File: rtengine/ahd_demosaic_RT.cc [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                            rix[0][2 - cng] = pix[0];
+                        }
+                        float xyz0 = 0.f;
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: Why no array here?

File: rtengine/ahd_demosaic_RT.cc [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                        lix = &lab[d][tr][tc];
+
+                        for (int i = 0; i < 4; i++) {
+                            ldiff[d][i] = std::fabs(lix[0][0] - lix[dir[i]][0]);
+                            abdiff[d][i] = SQR(lix[0][1] - lix[dir[i]][1])
+                                           + SQR(lix[0][2] - lix[dir[i]][2]);
+                        }
+                    }
+
+                    float leps = min(max(ldiff[0][0], ldiff[0][1]),
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: Here you use `rtengine::(min|max)()`, while above it's `std::(min|max)()`. There's a unification due, IMHO.

File: rtengine/ahd_demosaic_RT.cc [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                                     max(ldiff[1][2], ldiff[1][3]));
+                    float abeps = min(max(abdiff[0][0], abdiff[0][1]),
+                                      max(abdiff[1][2], abdiff[1][3]));
+
+                    for (int d = 0; d < 2; d++) {
+                        homo[d][tr][tc] = 0;
+                        for (int i = 0; i < 4; i++) {
+                            homo[d][tr][tc] += ((ldiff[d][i] <= leps) * (abdiff[d][i] <= abeps));
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: Redundant parentheses.

File: rtengine/ahd_demosaic_RT.cc [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+            if(plistener) {
+                progresscounter++;
+
+                if(progresscounter % 32 == 0) {
+#ifdef _OPENMP
+                    #pragma omp critical (ahdprogress)
+#endif
+                    {
+                        progress += (double)32 * ((TS - 32) * (TS - 32)) / (height * width);
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: Don't we have a `^2` function?

`(double)32` -> `32.0`.

File: rtengine/ahd_demosaic_RT.cc [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+                        progress = progress > 1.0 ? 1.0 : progress;
```
Comment (Floessie) [inline rtengine/ahd_demosaic_RT.cc]: Use `std::min()`.

File: rtengine/demosaic_algos.cc [hunk 11]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
...
 }
 
+void RawImageSource::border_interpolate(unsigned int border, float (*image)[3], unsigned int start, unsigned int end)
+{
+    unsigned row, col, y, x, f;
+    float sum[8];
+    unsigned int width = W, height = H;
+    unsigned int colors = 3;
+
+    if (end == 0 ) {
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: `if (!end) {` would express this nicer.

File: rtengine/demosaic_algos.cc [hunk 12]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
     free (image);
 }
 
+void RawImageSource::border_interpolate(unsigned int border, float (*image)[3], unsigned int start, unsigned int end)
+{
+    unsigned row, col, y, x, f;
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: C-style!

File: rtengine/demosaic_algos.cc [hunk 13]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
...
+    unsigned int width = W, height = H;
+    unsigned int colors = 3;
+
+    if (end == 0 ) {
+        end = H;
+    }
+
+    for (row = start; row < end; row++)
+        for (col = 0; col < width; col++) {
+            if (col == border && row >= border && row < height - border) {
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: Maybe there should be some checks for `border` relative to `start` and `height` when entering this function.

File: rtengine/demosaic_algos.cc [hunk 14]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
...
+                col = width - border;
+            }
+
+            memset (sum, 0, sizeof sum);
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: Super-ugly.

-> `sum = {}`

File: rtengine/demosaic_algos.cc [hunk 15]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
...
+
+            for (y = row - 1; y != row + 2; y++)
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: This assumes `start > 0`.

File: rtengine/demosaic_algos.cc [hunk 16]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
...
+                for (x = col - 1; x != col + 2; x++)
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: This assumes `col > 0`, and I can't immediately derive if that's true from the above.

File: rtengine/demosaic_algos.cc [hunk 17]
```diff
@@ -682,6 +682,41 @@ void RawImageSource::ppg_demosaic()
...
+                    if (y < height && x < width) {
+                        f = fc(y, x);
+                        sum[f] += image[y * width + x][f];
+                        sum[f + 4]++;
+                    }
+
+            f = fc(row, col);
+
+            FORCC if (c != f && sum[c + 4]) {
```
Comment (Floessie) [inline rtengine/demosaic_algos.cc]: Isn't `FORCC` dcraw specific?

**Review by Floessie — COMMENTED** [review]
Just a few comments when reading the whole patch at once.

HTH,
Flössie

File: rtengine/ahd_demosaic_RT.cc (hunk 1 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: @Floessie You're right. I simply copied that from amaze code. Will change that now.

File: rtengine/ahd_demosaic_RT.cc (hunk 2 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Done

File: rtengine/ahd_demosaic_RT.cc (hunk 3 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Done

File: rtengine/ahd_demosaic_RT.cc (hunk 5 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc:123]: @Floessie Do you mean the indexing by cng (ColourNonGreen)?

File: rtengine/ahd_demosaic_RT.cc (hunk 9 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Done

File: rtengine/ahd_demosaic_RT.cc (hunk 8 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Fixed

File: rtengine/ahd_demosaic_RT.cc (hunk 7 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Fixed

File: rtengine/demosaic_algos.cc (hunk 11 above)
Comment (heckflosse) [reply rtengine/demosaic_algos.cc]: I eliminated the whole function as it now was obsolete

File: rtengine/ahd_demosaic_RT.cc (hunk 6 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Changed

File: rtengine/ahd_demosaic_RT.cc (hunk 4 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Changed

File: rtengine/ahd_demosaic_RT.cc (hunk 10 above)
Comment (heckflosse) [reply rtengine/ahd_demosaic_RT.cc]: Done

**Pull request merged**